aws-crt-kotlin/native/src/aws/sdk/kotlin/crt/util/hashing/ShaNative.kt

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+import aws.sdk.kotlin.crt.Allocator
+import aws.sdk.kotlin.crt.CrtRuntimeException
+import aws.sdk.kotlin.crt.awsAssertOpSuccess
+import kotlinx.cinterop.*
+import libcrt.*
+
+/**
+ * A typealias to CRT's aws_<checksum_algorithm>_new functions (i.e aws_sha1_new, aws_sha256_new))
+ */
+internal typealias InitializeHashFn = (
+    allocator: CValuesRef<aws_allocator>?,
+) -> CPointer<aws_hash>?
+
+public open class ShaNative(
+    public val initializeFn: InitializeHashFn,
+) {
+    private var hash: CPointer<aws_hash>
+
+    init {
+        hash = initializeHash()
+    }
+
+    // aws_hash_update
+    public fun update(input: ByteArray, offset: Int, length: Int) {
+        val inputCursor = input.usePinned {
+            aws_byte_cursor_from_array(it.addressOf(offset), length.convert())
+        }
+        awsAssertOpSuccess(aws_hash_update(hash, inputCursor)) {
+            "aws_hash_update"
+        }
+    }
+
+    // aws_hash_finalize
+    public fun digest(): ByteArray {
+        val output: CPointer<aws_byte_buf> = aws_mem_calloc(
+            Allocator.Default,
+            1.convert(),
+            sizeOf<aws_byte_buf>().convert(),
+        )?.reinterpret() ?: throw CrtRuntimeException("aws_mem_calloc() aws_byte_buf")
```

Comment:
**Question**: What is this exception trying to say?

Reply:
It's referencing the name of the CRT function which failed (`aws_mem_calloc()`) specifically when trying to allocate memory for an `aws_byte_buf`.

We have a similar style across the rest of the Kotlin/Native implementation, it helps identify which C function failed.